### DISCUSSION TITLE
[Txt Data] Try to read txtData description of a server service

### DISCUSSION THREAD

**Discussion by estelleOl** [openthread/openthread#8250] [Q&A] [answered]
Hello,

I have created a client service with a txtData description but I try to read TxtData serve on server side. I do not have the txtData filled.

In client part here my code : 

 /* Get buffer instance name */
   serviceInst = otSrpClientBuffersGetServiceEntryInstanceNameString(entry, &size);
 
    strcpy(serviceInst, "board2Instance");
    
    /* Get buffer sub type labels array */
    subTypeArray = otSrpClientBuffersGetSubTypeLabelsArray(entry, &size);
    

    /* Set subType Label */
    subTypeArray[0] = "subtypeBoard2\0";
    
    /* Get service name buffer */
    serviceName = otSrpClientBuffersGetServiceEntryServiceNameString(entry, &size);

    strcpy(serviceName, "_subtypeBoard2._sub._testBoard2._udp");
    
    /* Get buffer txt entry */
    txtEntry = otSrpClientBuffersGetServiceEntryTxtBuffer(entry, &size);

    /* Set txtEntry */
    *txtEntry = 123;
    
    /* Set weight port priority service*/
    entry->mService.mPort = 111;
    entry->mService.mPriority = 1;
    entry->mService.mWeight = 1;
    
    /* Add a service to be registered with server */
    status = otSrpClientAddService(NULL, &entry->mService);

I can see that the TxtEntry is well set when I receive the srpClientCallback.

But when I try to use otSrpServerServiceGetTxtData on server side on this sqme service. I do have the txtEntry I have written.

Do I have misunderstood and the txtEntry is not the same as the txtData ?

**Comment by jwhui** [comment, marked as answer]
The length of the TXT entry value also needs to be set.

```
entry->mTxtEntry.mValueLength
```

The SRP Client CLI provides an example of configuring the TXT record:

https://github.com/openthread/openthread/blob/812c36f6581967b3d9af8361f00716c75f47473c/src/cli/cli_srp_client.cpp#L420-L428

**Reply by estelleOl** [reply]
I have modify my txt entry and put the lenght value but I do not succeed to register my service. The srpClientCallback as soon I put a lenght value return 6.
My code : 

```c++
strcpy(serviceInst, "board2Instance");

/* Get buffer sub type labels array */
subTypeArray = otSrpClientBuffersGetSubTypeLabelsArray(entry, &size);

/* Set subType Label */
subTypeArray[0] = "subtypeBoard2\0";

/* Get service name buffer */
serviceName = otSrpClientBuffersGetServiceEntryServiceNameString(entry, &size);

strcpy(serviceName, "_subtypeBoard2._sub._testBoard2._udp");

/* Get buffer txt entry */
txtEntry = otSrpClientBuffersGetServiceEntryTxtBuffer(entry, &size);

/* Set txtEntry */
  strcpy((char *)txtEntry, "testBoard2");
  entry->mTxtEntry.mValueLength = sizeof("testBoard2"); 

/* Set weight port priority service*/
entry->mService.mPort = 111;
entry->mService.mPriority = 1;
entry->mService.mWeight = 1;

/* Add a service to be registered with server */
status = otSrpClientAddService(NULL, &entry->mService);
```

If I do not put the length value I can register the service and the callback return no error but still the same problem srp server do not see the txt data with the method : 

 uint16_t  aDataLenght;
 const uint8_t * txtData = OtSrpServerServiceGetTxtData(pService, &aDataLenght);

dataLenght return 1 and txtData is empty.

Is txtEntry is the srp client input to set txtData in srp server side ?

**Reply by abtink** [reply]
The `mTxtEntries` that you include  in the `otSrpClientService` must be an array of _valid_ `otDnsTxtEntry` items.

Please check the documentation of `otDnsTxtEntry` (and/or RFC 6763):
```c++
/**
 * This structure represents a TXT record entry representing a key/value pair (RFC 6763 - section 6.3).
 *
 * The string buffers pointed to by `mKey` and `mValue` MUST persist and remain unchanged after an instance of such
 * structure is passed to OpenThread (as part of `otSrpClientService` instance).
 *
 * An array of `otDnsTxtEntry` entries are used in `otSrpClientService` to specify the full TXT record (a list of
 * entries).
 *
 */
typedef struct otDnsTxtEntry
{
    /**
     * The TXT record key string.
     *
     * If `mKey` is not NULL, then it MUST be a null-terminated C string. The entry is treated as key/value pair with
     * `mValue` buffer providing the value.
     *   - The entry is encoded as follows:
     *        - A single string length byte followed by "key=value" format (without the quotation marks).
              - In this case, the overall encoded length must be 255 bytes or less.
     *   - If `mValue` is NULL, then key is treated as a boolean attribute and encoded as "key" (with no `=`).
     *   - If `mValue` is not NULL but `mValueLength` is zero, then it is treated as empty value and encoded as "key=".
     *
     * If `mKey` is NULL, then `mValue` buffer is treated as an already encoded TXT-DATA and is appended as is in the
     * DNS message.
     *
     */
    const char *   mKey;
    const uint8_t *mValue;       ///< The TXT record value or already encoded TXT-DATA (depending on `mKey`).
    uint16_t       mValueLength; ///< Number of bytes in `mValue` buffer.
} otDnsTxtEntry;
```

You can provide a bunch of "key/value" pairs and then OT implementation will encode it and construct the TXT-Data.
Of if `mKey` is NULL, then caller must already provide the encoded TXT-DATA (which should follow format as in RFC 6763).

In your code, you seem to leave `mKey` as NULL, then copy `"testBoard2"` into the buffer which does not follow proper format.

**Reply by estelleOl** [reply]
Thank you , your response has resolved my issue
